planner/core/task.go

```diff
@@ -1302,7 +1302,31 @@ func (p *PhysicalProjection) attach2Task(tasks ...task) task {
 	return t
 }
 
+func (p *PhysicalUnionAll) attach2MppTasks(tasks ...task) task {
+	t := &mppTask{p: p}
+	childPlans := make([]PhysicalPlan, 0, len(tasks))
+	var childMaxCost float64
+	for _, tk := range tasks {
+		if mpp, ok := tk.(*mppTask); ok && !tk.invalid() {
+			childCost := mpp.cost()
+			if childCost > childMaxCost {
+				childMaxCost = childCost
+			}
+			childPlans = append(childPlans, mpp.plan())
+		} else {
+			return invalidTask
+		}
+	}
+	p.SetChildren(childPlans...)
+	t.cst = childMaxCost
+	p.cost = t.cost()
+	return t
+}
+
 func (p *PhysicalUnionAll) attach2Task(tasks ...task) task {
+	if _, ok := tasks[0].(*mppTask); ok {
```

Comment:
even if their children are mpp tasks, the unionAll may also return root task

Reply:
in what case?

Reply:
in `TiDBMPP` mode, union all is at tidb, then its children run in MPP.